Project 4


In [1]:
#importing the required libraries
import matplotlib.pyplot as plt
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn import metrics
from sklearn import model_selection
from sklearn.preprocessing import StandardScaler
from sklearn.tree import plot_tree

In [10]:
seed = 520
np.set_printoptions(precision=3)

In [11]:
#loading the excel file AndroidmalwareSmall.xlsx
android_df=pd.read_excel("AndroidmalwareSmall.xlsx")

In [12]:
#printing 5 rows of the dataset
android_df.head()

android.permission.GET_ACCOUNTS  \
0                                0   
1                                1   
2                                0   
3                                0   
4                                0   

   com.sonyericsson.home.permission.BROADCAST_BADGE  \
0                                                 0   
1                                                 0   
2                                                 0   
3                                                 0   
4                                                 0   

   android.permission.READ_PROFILE  android.permission.MANAGE_ACCOUNTS  \
0                                0                                   0   
1                                0                                   0   
2                                0                                   0   
3                                0                                   0   
4                                0                                   0   

   android.permission.WRITE_SYNC_SETTINGS  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   android.permission.READ_EXTERNAL_STORAGE  android.permission.RECEIVE_SMS  \
0                                         0                               0   
1                                         0                               0   
2                                         1                               0   
3                                         1                               0   
4                                         0                               0   

   com.android.launcher.permission.READ_SETTINGS  \
0                                              0   
1                                              0   
2                                              0   
3                                              0   
4                                              0   

   android.permission.WRITE_SETTINGS  \
0                                  0   
1                                  0   
2                                  0   
3                                  1   
4                                  0   

   com.google.android.providers.gsf.permission.READ_GSERVICES  ...  \
0                                                  0           ...   
1                                                  0           ...   
2                                                  0           ...   
3                                                  0           ...   
4                                                  0           ...   

   com.android.launcher.permission.UNINSTALL_SHORTCUT  \
0                                                  0    
1                                                  0    
2                                                  0    
3                                                  0    
4                                                  0    

   com.sec.android.iap.permission.BILLING  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   com.htc.launcher.permission.UPDATE_SHORTCUT  \
0                                            0   
1                                            0   
2                                            0   
3                                            0   
4                                            0   

   com.sec.android.provider.badge.permission.WRITE  \
0                                                0   
1                                                0   
2                                                0   
3                                                0   
4                                                0   

   android.permission.ACCESS_NETWORK_

In [13]:
#splitting the data into train and test as 80% and 20%
X=android_df.iloc[:,1:87]
y=android_df['Result']
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test= train_test_split(X,y,test_size=0.2,random_state=1)


In [16]:
#printing the dimensions of X's and y's trains and tests 
print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(8000, 86)
(2000, 86)
(8000,)
(2000,)


RandomForest

In [2]:
#defining a function for scaling the data
def create_scaler_standard():
    return StandardScaler()

create_scaler = create_scaler_standard

In [8]:
#defining a function for randomforestclassifier model
def create_model_random_forest():
    model_rf = RandomForestClassifier(n_estimators=100, 
                                   min_samples_split=5, 
                                   random_state=seed, 
                                   n_jobs=-1)
    return model_rf

create_model_rf = create_model_random_forest

In [17]:
#scaling the model and fitting the model 
scaler = create_scaler()
model_rf = create_model_rf()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)
model_rf.fit(X_train_scaled, y_train)

RandomForestClassifier(min_samples_split=5, n_jobs=-1, random_state=520)

In [18]:
#Predicting the model for train and test
y_train_pred = model_rf.predict(X_train_scaled)
y_test_pred = model_rf.predict(X_test_scaled)

In [19]:
#print the accuracy,classification report and confusion matrix for the model
print('Model evaluation (train):')
print('Accuracy:')
print(metrics.accuracy_score(y_train, y_train_pred))
print("Classification report:")
print(metrics.classification_report(y_train, y_train_pred))
print("Confusion matrix (train):")
print(metrics.confusion_matrix(y_train, y_train_pred))

Model evaluation (train):
Accuracy:
1.0
Classification report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      4043
           1       1.00      1.00      1.00      3957

    accuracy                           1.00      8000
   macro avg       1.00      1.00      1.00      8000
weighted avg       1.00      1.00      1.00      8000

Confusion matrix (train):
[[4043    0]
 [   0 3957]]


Got Accuracy of 100% for RandomForest model.

Hyperparameter Tuning for RandomForestClassifier

In [26]:
#Using gridsearchcv to tuning the model
#training and fitting the model
model = model_selection.GridSearchCV(create_model_rf(),
                                     cv=5,
                                     param_grid={
                                         "max_depth": list(range(1, 40, 2)),
                                         "min_samples_split": list(range(2, 5, 
2))
                                      })
model.fit(X_train_scaled, y_train)


GridSearchCV(cv=5,
             estimator=RandomForestClassifier(min_samples_split=5, n_jobs=-1,
                                              random_state=520),
             param_grid={'max_depth': [1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 21,
                                       23, 25, 27, 29, 31, 33, 35, 37, 39],
                         'min_samples_split': [2, 4]})

In [46]:
#print the accuracy,classification report and confusion matrix for the model
#After hyperparameter tuning
print('Model evaluation (Optimal Hyperparameters):')
print('Optimal parameters:', model.best_params_)
y_test_hat = model.predict(X_test_scaled)
print('Accuracy:')
print(metrics.accuracy_score(y_test, y_test_hat))
print("Classification report:")
print(metrics.classification_report(y_test, y_test_hat))
print("Confusion matrix (Optimal Hyperparameters):")
print(metrics.confusion_matrix(y_test, y_test_hat))

Model evaluation (Optimal Hyperparameters):
Optimal parameters: {'max_depth': 7, 'min_samples_split': 2}
Accuracy:
1.0
Classification report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1001
           1       1.00      1.00      1.00       999

    accuracy                           1.00      2000
   macro avg       1.00      1.00      1.00      2000
weighted avg       1.00      1.00      1.00      2000

Confusion matrix (Optimal Hyperparameters):
[[1001    0]
 [   0  999]]


There is no change in accuracy.

AdaBoost

In [35]:
#create a model for adaboostclassifier with decisiontree as base estimator
param_grid = {"base_estimator__criterion" : ["gini", "entropy"],
              "base_estimator__splitter" :   ["best", "random"],
              "n_estimators": [1, 2]
             }
Decision_TC = DecisionTreeClassifier(random_state = 42, max_features = "auto", class_weight = "balanced",max_depth = None)

model_abc = AdaBoostClassifier(base_estimator = Decision_TC)


In [36]:
#training and fitting the model
scaler.fit(X_train)
X_trainscaled = scaler.transform(X_train)
X_testscaled = scaler.transform(X_test)
model_abc.fit(X_trainscaled, y_train)

AdaBoostClassifier(base_estimator=DecisionTreeClassifier(class_weight='balanced',
                                                         max_features='auto',
                                                         random_state=42))

In [38]:
#predicting the model for train and test
y_trainpred = model_abc.predict(X_trainscaled)
y_testpred = model_abc.predict(X_testscaled)

In [39]:
#print the accuracy,classification report and confusion matrix for the model
print('Model evaluation (train):')
print('Accuracy:')
print(metrics.accuracy_score(y_train, y_trainpred))
print("Classification report:")
print(metrics.classification_report(y_train, y_trainpred))
print("Confusion matrix (train):")
print(metrics.confusion_matrix(y_train, y_trainpred))

Model evaluation (train):
Accuracy:
1.0
Classification report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      4043
           1       1.00      1.00      1.00      3957

    accuracy                           1.00      8000
   macro avg       1.00      1.00      1.00      8000
weighted avg       1.00      1.00      1.00      8000

Confusion matrix (train):
[[4043    0]
 [   0 3957]]


Got Accuracy of 100% for AdaBoost model

In [42]:
#Using gridsearchcv to tuning the model
#training and fitting the model
model_a = model_selection.GridSearchCV(model_abc,cv=5,param_grid=param_grid)
model_a.fit(X_trainscaled, y_train)

GridSearchCV(cv=5,
             estimator=AdaBoostClassifier(base_estimator=DecisionTreeClassifier(class_weight='balanced',
                                                                                max_features='auto',
                                                                                random_state=42)),
             param_grid={'base_estimator__criterion': ['gini', 'entropy'],
                         'base_estimator__splitter': ['best', 'random'],
                         'n_estimators': [1, 2]})

HyperParameter tuning for AdaBoost

In [45]:
#print the accuracy,classification report and confusion matrix for the model
#After hyperparameter tuning
print('Model evaluation (Optimal Hyperparameters):')
print('Optimal parameters:', model_a.best_params_)
y_testhat = model_a.predict(X_testscaled)
print('Accuracy:')
print(metrics.accuracy_score(y_test, y_testhat))
print("Classification report:")
print(metrics.classification_report(y_test, y_testhat))
print("Confusion matrix (Optimal Hyperparameters):")
print(metrics.confusion_matrix(y_test, y_testhat))

Model evaluation (Optimal Hyperparameters):
Optimal parameters: {'base_estimator__criterion': 'entropy', 'base_estimator__splitter': 'random', 'n_estimators': 1}
Accuracy:
0.987
Classification report:
              precision    recall  f1-score   support

           0       0.99      0.98      0.99      1001
           1       0.98      0.99      0.99       999

    accuracy                           0.99      2000
   macro avg       0.99      0.99      0.99      2000
weighted avg       0.99      0.99      0.99      2000

Confusion matrix (Optimal Hyperparameters):
[[984  17]
 [  9 990]]


The Accuracy for RandomForestClassifier model as well as AdaBoostclassifier model with DecisionTreeClassifier as base learner is 100%.But after performing hyperparameter tuning, RandomForestclassifier is the best model since it has 100% accuracy when compared to AbaBoostClassifier which is of 98.7%.